src/lambertw.jl

```diff
@@ -0,0 +1,386 @@
+import Base: convert
+#export lambertw, lambertwbp
+using Compat
+
+const euler =
+if isdefined(Base, :MathConstants)
+    Base.MathConstants.e
+else
+    e
+end
+
+const omega_const_bf_ = Ref{BigFloat}()
+
+function __init__()
+    omega_const_bf_[] =
+        parse(BigFloat,"0.5671432904097838729999686622103555497538157871865125081351310792230457930866845666932194")
+end
+
+#### Lambert W function ####
+
+const LAMBERTW_USE_NAN = false
+
+macro baddomain(v)
+    if LAMBERTW_USE_NAN
+        return :(return(NaN))
+    else
+        return esc(:(throw(DomainError($v))))
+    end
+end
+
+# Use Halley's root-finding method to find x = lambertw(z) with
+# initial point x.
+function _lambertw(z::T, x::T) where T <: Number
+    two_t = convert(T,2)
+    lastx = x
+    lastdiff = zero(T)
+    for i in 1:100
+        ex = exp(x)
+        xexz = x * ex - z
+        x1 = x + 1
+        x = x - xexz / (ex * x1 - (x + two_t) * xexz / (two_t * x1 ) )
+        xdiff = abs(lastx - x)
+        xdiff <= 2*eps(abs(lastx)) && break
+        lastdiff == diff && break
+        lastx = x
+        lastdiff = xdiff
+    end
+    x
+end
+
+### Real z ###
+
+# Real x, k = 0
+
+# The fancy initial condition selection does not seem to help speed, but we leave it for now.
+function lambertwk0(x::T)::T where T<:AbstractFloat
+    x == Inf && return Inf
```

Comment:
Perhaps `isfinite(x) || return x`? That will also handle the case where `x` is `NaN`, or if `T` is anything other than `Float64` (`Inf` is a `Float64` literal).

Reply:
I don't recall my reasoning here. It is highly likely that I compared to python and Mathematica.  But, maybe there is a Julia convention. Some clues for what to do:
```
julia> sin(NaN)
NaN

julia> sin(Inf)
ERROR: DomainError with Inf:
sin(x) is only defined for finite x.
```
But `sin` does not tend to a limit with increasing `x`.  So `Inf` is in no sense the correct result.  `lambertw` tends to infinity, so `Inf` in this sense is the correct answer.
```
In [5]: lambertw(inf)  # python mpmath
Out[5]: mpf('+inf')
In [9]: lambertw(nan)
Out[9]: mpf('nan')
```
Another issue is:

```
julia> convert(BigFloat,Inf)
Inf

julia> typeof(convert(BigFloat,Inf))
BigFloat
```
Yikes, no visible difference.  In any case, at a minimum, this should be `convert(T,Inf)`.